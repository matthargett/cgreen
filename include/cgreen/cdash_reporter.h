#ifndef CDASH_REPORTER_HEADER
#define CDASH_REPORTER_HEADER

#include <cgreen/reporter.h>
#include <stdio.h>

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

typedef struct CDashInfo_ CDashInfo;
struct CDashInfo_ {
	char *name;
	char *build;
	char *type;
	char *hostname;
	char *os_name;
	char *os_platform;
	char *os_release;
	char *os_version;
};

typedef int CDashPrinter(FILE *, const char *format, ...);

extern void set_cdash_printer(TestReporter *reporter, CDashPrinter *printer);
extern TestReporter *create_cdash_reporter(CDashInfo *cdash);

#ifdef __cplusplus
    }
}
#endif

#endif
