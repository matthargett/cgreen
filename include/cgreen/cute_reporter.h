#ifndef CUTE_REPORTER_HEADER
#define CUTE_REPORTER_HEADER

#include <cgreen/reporter.h>

#ifdef __cplusplus
namespace cgreen {
    extern "C" {
#endif

typedef int CutePrinter(const char *format, ...);


extern TestReporter *create_cute_reporter(void);
extern void set_cute_printer(TestReporter *reporter, CutePrinter *printer);

#ifdef __cplusplus
    }
}
#endif

#endif
