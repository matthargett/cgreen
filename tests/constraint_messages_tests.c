/* Note: the whole point of these tests is that they should fail!
   This makes it hard to think straight and to give the variables good
   names ;-) */

#include <cgreen/cgreen.h>
#include <cgreen/mocks.h>
#include <signal.h>
#include <float.h>

#ifdef __cplusplus
using namespace cgreen;
#endif

Describe(ConstraintMessage);
BeforeEach(ConstraintMessage) {}
AfterEach(ConstraintMessage) {}

Ensure(ConstraintMessage,for_is_null) {
    int i;
    int *pointer = &i;
    assert_that(pointer, is_null);
}

Ensure(ConstraintMessage,for_is_non_null) {
    int *pointer = NULL;
    assert_that(pointer, is_non_null);
}

Ensure(ConstraintMessage,for_is_equal_to) {
    int fourty_five = 45, thirty_three = 33;
    assert_that(fourty_five, is_equal_to(thirty_three));
}

Ensure(ConstraintMessage, for_is_equal_to_hex) {
    char chars[4];
    memset(chars, 0xaa, sizeof(chars));
    assert_that(chars[0], is_equal_to_hex(0xbb));
}

Ensure(ConstraintMessage,for_is_not_equal_to) {
    int should_not_be_fourty_five = 45, forty_five = 45;
    assert_that(should_not_be_fourty_five, is_not_equal_to(forty_five));
}

Ensure(ConstraintMessage,for_is_greater_than) {
    int fourty_five = 45, thirty_three = 33;
    assert_that(thirty_three, is_greater_than(fourty_five));
}

Ensure(ConstraintMessage,for_is_less_than) {
    int fourty_five = 45, thirty_three = 33;
    assert_that(fourty_five, is_less_than(thirty_three));
}

Ensure(ConstraintMessage,for_is_equal_to_double) {
    double four_point_five = 4.5f, three_point_three = 3.3f;
    assert_that_double(four_point_five, is_equal_to_double(three_point_three));
}

Ensure(ConstraintMessage,for_is_not_equal_to_double) {
    significant_figures_for_assert_double_are(4);
    double epsilon = 1.0e-4;
    double four_point_five = 4.5, almost_four_point_five = 4.5 - epsilon;
    assert_that_double(four_point_five, is_not_equal_to_double(almost_four_point_five));
}

Ensure(ConstraintMessage,for_is_greater_than_double) {
    double four_point_five = 4.5f, three_point_three = 3.3f;
    assert_that_double(three_point_three, is_greater_than_double(four_point_five));
}

Ensure(ConstraintMessage,for_is_less_than_double) {
    double four_point_five = 4.5f, three_point_three = 3.3f;
    assert_that_double(four_point_five, is_less_than_double(three_point_three));
}

Ensure(ConstraintMessage,for_is_equal_to_contents_of) {
    int fourty_five[45] = {45, 44, 43}, thirty_three[33] = {45, 44, 33};
    assert_that(thirty_three, is_equal_to_contents_of(fourty_five, 55));
}

Ensure(ConstraintMessage,for_is_not_equal_to_contents_of) {
    int fourty_five_and_up[3] = {45, 46, 47}, another_fourty_five_and_up[3] = {45, 46, 47};
    assert_that(fourty_five_and_up, is_not_equal_to_contents_of(another_fourty_five_and_up, 3));
}

Ensure(ConstraintMessage,for_is_equal_to_string) {
    const char *fourty_five = "this string is fortyfive", *thirty_three = "this string is thirtythree";
    assert_that(thirty_three, is_equal_to_string(fourty_five));
}

Ensure(ConstraintMessage,for_is_not_equal_to_string) {
    const char *fourty_five = "this string is fourtyfive", *another_fourty_five = "this string is fourtyfive";
    assert_that(another_fourty_five, is_not_equal_to_string(fourty_five));
}

Ensure(ConstraintMessage,for_contains_string) {
    const char *fourty_five = "fortyfive", *not_containing_fourty_five = "this text is thirtythree";
    assert_that(not_containing_fourty_five, contains_string(fourty_five));
}

Ensure(ConstraintMessage,for_does_not_contain_string) {
    const char *contains_fourty_five = "this string is fourtyfive", *fourty_five = "fourtyfive";
    assert_that(contains_fourty_five, does_not_contain_string(fourty_five));
}

Ensure(ConstraintMessage,for_begins_with_string) {
    const char *does_not_begin_with_fourty_five = "this string does not begin with fortyfive", *fourty_five = "fourtyfive";
    assert_that(does_not_begin_with_fourty_five, begins_with_string(fourty_five));
}

Ensure(ConstraintMessage,for_equal_to_double) {
  assert_that_double(0, is_equal_to_double(1));
}

Ensure(ConstraintMessage,for_equal_to_double_negative) {
  assert_that_double(-1, is_equal_to_double(-2));
}

Ensure(ConstraintMessage,for_not_equal_to_double) {
  assert_that_double(0, is_not_equal_to_double(0));
}

Ensure(ConstraintMessage,for_not_equal_to_double_negative) {
  assert_that_double(-1, is_not_equal_to_double(-1));
}

Ensure(ConstraintMessage,for_is_less_than_double_with_accuracy) {
  significant_figures_for_assert_double_are(4);
  assert_that_double(1.0, is_less_than_double(1.0 - 1.0e-3 - DBL_EPSILON));
}

Ensure(ConstraintMessage,for_is_greater_than_double_with_accuracy) {
  significant_figures_for_assert_double_are(4);
  assert_that_double(1.0, is_greater_than_double(1.0 + 1.0e-3 + DBL_EPSILON));
}

Ensure(ConstraintMessage,for_assert_that) {
    assert_that(0 == 1);
}

void some_mock(int parameter) {
    mock(parameter);
}

Ensure(ConstraintMessage, for_mock_called_without_expectation) {
    some_mock(0);
}

Ensure(ConstraintMessage, for_mock_called_more_times_than_expected) {
    expect(some_mock);
    some_mock(0);
    some_mock(0);
}

Ensure(ConstraintMessage, for_mock_called_with_unexpected_parameter_value) {
    expect(some_mock, when(parameter, is_equal_to(1)));
    some_mock(0);
}

/* And here are some tests that will fail with "illegal" type of messages */
Ensure(ConstraintMessage, for_always_followed_by_expectation) {
    always_expect(some_mock, when(parameter, is_equal_to(1)));
    expect(some_mock, when(parameter, is_equal_to(0)));
}

Ensure(ConstraintMessage, for_violated_never_expect) {
    never_expect(some_mock, when(parameter, is_equal_to(1)));
    some_mock(1);
}

Ensure(ConstraintMessage, for_mock_parameter_name_not_matching_constraint_parameter_name) {
    expect(some_mock, when(PARAMETER, is_equal_to(0)));
    some_mock(0);
}

void forgot_to_pass_parameters_mock(int x) {
    (void)x; // to silence compiler warning
    mock();
}

Ensure(ConstraintMessage, for_no_mock_parameters_with_parameter_constraint) {
    expect(forgot_to_pass_parameters_mock, when(x, is_equal_to(0)));
    forgot_to_pass_parameters_mock(0);
}

Ensure(ConstraintMessage, increments_exception_count_when_terminating_via_SIGQUIT) {
    raise(SIGQUIT);
}

Ensure(ConstraintMessage, increments_exception_count_when_terminating_via_SIGTERM) {
    raise(SIGTERM);
}

#ifdef __cplusplus
Ensure(ConstraintMessage, for_incorrect_assert_throws) {
    assert_throws(std::string, throw "something else");
}

Ensure(ConstraintMessage, for_exception_was_expected_but_nothing_thrown) {
    assert_throws(std::string, (void)5);
}
#endif
