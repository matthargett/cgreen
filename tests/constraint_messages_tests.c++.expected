Running "constraint_messages_tests" (36 tests)...
constraint_messages_tests.c: Failure: ConstraintMessage -> for_always_followed_by_expectation 
	Mocked function [some_mock] already has an expectation that it will always be called a certain way; any expectations declared after an always expectation are invalid

constraint_messages_tests.c: Failure: ConstraintMessage -> for_assert_that 
	Expected [0 == 1] to [be true]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_begins_with_string 
	Expected [does_not_begin_with_fourty_five] to [begin with string] [fourty_five]
		actual value:			["this string does not begin with fortyfive"]
		expected to begin with:		["fourtyfive"]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_contains_string 
	Expected [not_containing_fourty_five] to [contain string] [fourty_five]
		actual value:			["this text is thirtythree"]
		expected to contain:		["fortyfive"]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_does_not_contain_string 
	Expected [contains_fourty_five] to [not contain string] [fourty_five]
		actual value:			["this string is fourtyfive"]
		expected to not contain:	["fourtyfive"]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_equal_to_double 
	Expected [0] to [equal double] [1] within [8] significant figures
		actual value:	0.000000
		expected value:	1.000000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_equal_to_double_negative 
	Expected [-1] to [equal double] [-2] within [8] significant figures
		actual value:	-1.000000
		expected value:	-2.000000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_exception_was_expected_but_nothing_thrown 
	Expected [(void)5] to throw [std::string]

constraint_messages_tests.c: Exception: ConstraintMessage -> for_incorrect_assert_throws 
	an exception was thrown during test: [something else]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_equal_to 
	Expected [fourty_five] to [equal] [thirty_three]
		actual value:			[45]
		expected value:			[33]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_equal_to_contents_of 
	Expected [thirty_three] to [equal contents of] [fourty_five]
		at offset:			[9]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_equal_to_double 
	Expected [four_point_five] to [equal double] [three_point_three] within [8] significant figures
		actual value:	4.500000
		expected value:	3.300000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_equal_to_hex 
	Expected [chars[0]] to [equal] [0xbb]
		actual value:			[0xffffffaa]
		expected value:			[0xbb]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_equal_to_string 
	Expected [thirty_three] to [equal string] [fourty_five]
		actual value:			["this string is thirtythree"]
		expected to equal:		["this string is fortyfive"]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_greater_than 
	Expected [thirty_three] to [be greater than] [fourty_five]
		actual value:			[33]
		expected to be greater than:	[45]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_greater_than_double 
	Expected [three_point_three] to [be greater than double] [four_point_five] within [8] significant figures
		actual value:	3.300000
		expected value:	4.500000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_greater_than_double_with_accuracy 
	Expected [1.0] to [be greater than double] [1.0 + 1.0e-3 + DBL_EPSILON] within [4] significant figures
		actual value:	1.000000
		expected value:	1.001000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_less_than 
	Expected [fourty_five] to [be less than] [thirty_three]
		actual value:			[45]
		expected to be less than:	[33]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_less_than_double 
	Expected [four_point_five] to [be less than double] [three_point_three] within [8] significant figures
		actual value:	4.500000
		expected value:	3.300000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_less_than_double_with_accuracy 
	Expected [1.0] to [be less than double] [1.0 - 1.0e-3 - DBL_EPSILON] within [4] significant figures
		actual value:	1.000000
		expected value:	0.999000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_non_null 
	Expected [pointer] to [be non null]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_not_equal_to 
	Expected [should_not_be_fourty_five] to [not equal] [forty_five]
		actual value:			[45]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_not_equal_to_contents_of 
	Expected [fourty_five_and_up] to [not equal contents of] [another_fourty_five_and_up]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_not_equal_to_double 
	Expected [four_point_five] to [not equal double] [almost_four_point_five] within [4] significant figures
		actual value:	4.500000
		expected value:	4.499900

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_not_equal_to_string 
	Expected [another_fourty_five] to [not equal string] [fourty_five]
		actual value:			["this string is fourtyfive"]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_is_null 
	Expected [pointer] to [be null]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_mock_called_more_times_than_expected 
	Mocked function [some_mock] was called too many times

constraint_messages_tests.c: Failure: ConstraintMessage -> for_mock_called_with_unexpected_parameter_value 
	Expected [[parameter] parameter in [some_mock]] to [equal] [1]
		actual value:			[0]
		expected value:			[1]

constraint_messages_tests.c: Failure: ConstraintMessage -> for_mock_called_without_expectation 
	Mocked function [some_mock] did not have an expectation that it would be called

constraint_messages_tests.c: Failure: ConstraintMessage -> for_mock_parameter_name_not_matching_constraint_parameter_name 
	Mocked function [some_mock] did not define a parameter named [PARAMETER]. Did you misspell it in the expectation or forget it in the mock's argument list?

constraint_messages_tests.c: Failure: ConstraintMessage -> for_no_mock_parameters_with_parameter_constraint 
	Mocked function [forgot_to_pass_parameters_mock] did not define a parameter named [x]. Did you misspell it in the expectation or forget it in the mock's argument list?

constraint_messages_tests.c: Failure: ConstraintMessage -> for_not_equal_to_double 
	Expected [0] to [not equal double] [0] within [8] significant figures
		actual value:	0.000000
		expected value:	0.000000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_not_equal_to_double_negative 
	Expected [-1] to [not equal double] [-1] within [8] significant figures
		actual value:	-1.000000
		expected value:	-1.000000

constraint_messages_tests.c: Failure: ConstraintMessage -> for_violated_never_expect 
	Mocked function [some_mock] has an expectation that it will never be called, but it was

constraint_messages_tests.c: Exception: ConstraintMessage -> increments_exception_count_when_terminating_via_SIGQUIT 
	Test terminated with signal: Quit

constraint_messages_tests.c: Exception: ConstraintMessage -> increments_exception_count_when_terminating_via_SIGTERM 
	Test terminated with signal: Terminated

Completed "ConstraintMessage": 0 passes, 33 failures, 3 exceptions in 0ms.
Completed "constraint_messages_tests": 0 passes, 33 failures, 3 exceptions in 0ms.
